### PR TITLE
storage: delete TxnStore, migrate tests.

### PR DESCRIPTION
**Pull request by disksing**
1) The new Scheduler does not use `TxnStore` any more, we need to delete it.
2) There were many tests depend on `TxnStore`, so `SyncStorage` (which wraps `Storage` with sync API) is introduced to migrate tests.
3) In order to make 2) easier,  `async!` and `await!` are defined.

P.S. The PR looks huge, but the main part is moving tests from `store.rs` to `sync_storage.rs`. I guess it will not be a disaster to review :)

PTAL @zhangjinpeng1987 @BusyJay  @ngaut

### CODE REVIEW

File: src/storage/engine/mod.rs [hunk 1]
```diff
@@ -48,27 +47,21 @@ pub trait Engine: Send + Sync + Debug {
...
 
     fn write(&self, ctx: &Context, batch: Vec<Modify>) -> Result<()> {
-        let finished = Event::new();
-        let finished2 = finished.clone();
         let timeout = Duration::from_secs(DEFAULT_TIMEOUT_SECS);
-
-        try!(self.async_write(ctx, batch, box move |res| finished2.set(res)));
-        if finished.wait_timeout(Some(timeout)) {
-            return finished.take().unwrap();
+        let event = async!(_cb, try!(self.async_write(ctx, batch, _cb)));
```
Comment (BusyJay) [inline src/storage/engine/mod.rs]: I think `async!(|cb| try!(self.async_write(ctx, batch, cb)))` is more readable.

Comment (disksing) [inline src/storage/engine/mod.rs]: @BusyJay Good idea! I also felt format is a little ugly, but was not able to come up with the better one.

File: src/util/event.rs [hunk 2]
```diff
@@ -154,6 +154,45 @@ impl<T> Drop for Event<T> {
...
+            $expr;
+            e1
+        }
+    }
+}
+
+/// `await!` waits on an Event (usually created by `async!`) and returns
+/// Option<Res>.
+#[macro_export]
+macro_rules! await{
```
Comment (zhangjinpeng1987) [inline src/util/event.rs]: What's await means?

Comment (disksing) [inline src/util/event.rs]: _async/await_ comes from C#. See https://msdn.microsoft.com/en-us/library/mt674882.aspx
I think the asynchronous model we use is alike, so inherit the names.

File: src/util/event.rs [hunk 3]
```diff
@@ -154,6 +154,45 @@ impl<T> Drop for Event<T> {
...
+    ($event:ident) => {
+        {
+            if $event.wait_timeout(None) {
```
Comment (BusyJay) [inline src/util/event.rs]: Can this be extracted to a local function?

**Comment by BusyJay** [comment]
It seems more appropriate to move `SyncStorage` to tests directory.

**Comment by disksing** [comment]
PTAL @BusyJay @ngaut

File: src/util/event.rs [hunk 4]
```diff
@@ -154,39 +154,31 @@ impl<T> Drop for Event<T> {
...
-/// `$expr`.
+/// `wait_event!` waits for async expression. It returns `Option<Res>`
+/// after the expression get executed.
 #[macro_export]
-macro_rules! async {
-    (|$cb:ident| $expr:expr) => {
+macro_rules! wait_event {
+    ($expr:expr) => {
         {
             let e1 = $crate::util::event::Event::new();
```
Comment (BusyJay) [inline src/util/event.rs]: This code is duplicated with L176，may be better to be extracted to a function.

File: src/storage/engine/mod.rs [hunk 5]
```diff
@@ -48,27 +47,19 @@ pub trait Engine: Send + Sync + Debug {
...
 
     fn write(&self, ctx: &Context, batch: Vec<Modify>) -> Result<()> {
-        let finished = Event::new();
-        let finished2 = finished.clone();
         let timeout = Duration::from_secs(DEFAULT_TIMEOUT_SECS);
-
-        try!(self.async_write(ctx, batch, box move |res| finished2.set(res)));
-        if finished.wait_timeout(Some(timeout)) {
-            return finished.take().unwrap();
+        match wait_event!(|cb| self.async_write(ctx, batch, cb).unwrap(), timeout) {
```
Comment (BusyJay) [inline src/storage/engine/mod.rs]: ok_or_else

**Comment by BusyJay** [comment]
rest LGTM

File: src/util/event.rs [hunk 6]
```diff
@@ -159,25 +159,18 @@ impl<T> Drop for Event<T> {
...
-            } else {
-                None
-            }
-        }
+        wait_event!(IMPL $expr, None)
     };
     ($expr:expr, $timeout:expr) => {
+        wait_event!(IMPL $expr, Some($timeout))
+    };
+    (IMPL $expr:expr, $timeout:expr) => {
```
Comment (BusyJay) [inline src/util/event.rs]: Can we define a new macro instead? This IMPL looks tricky and ugly.

**Comment by BusyJay** [comment]
LGTM

File: src/util/event.rs [hunk 7]
```diff
@@ -154,6 +154,32 @@ impl<T> Drop for Event<T> {
     }
 }
 
+/// `wait_event!` waits for async expression. It returns `Option<Res>`
+/// after the expression get executed.
+#[macro_export]
+macro_rules! wait_event {
+    ($expr:expr) => {
+        wait_event!(IMPL $expr, None)
```
Comment (siddontang) [inline src/util/event.rs]: what does IMPL mean?

Comment (disksing) [inline src/util/event.rs]: Just a flag to let the macro call itself recursively.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**